cmd/skuba/cluster/check.go

```diff
@@ -0,0 +1,108 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package cluster
+
+import (
+	"fmt"
+	"io/ioutil"
+
+	"github.com/SUSE/skuba/pkg/skuba"
+	"github.com/rikatz/kubepug/lib"
+	"github.com/rikatz/kubepug/pkg/formatter"
+	"github.com/spf13/cobra"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/klog"
+)
+
+type checkOptions struct {
+	K8sVersion      string
+	ForceDownload   bool
+	APIWalk         bool
+	SwaggerDir      string
+	ShowDescription bool
+}
+
+var (
+	format    string
+	filename  string
+	inputFile string
+)
+
+// NewCheckCmd creates a new `skuba check` cobra command
+func NewCheckCmd() *cobra.Command {
+	checkOptions := &checkOptions{}
+
+	cmd := &cobra.Command{
+		Use:   "check k8s-version=<version> swaggerDir=<directory> --api-walk=<true|fasle>",
+		Short: "Print Check information",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			kubernetesConfigFlags := genericclioptions.NewConfigFlags(true)
+			kubeAdminFile := skuba.KubeConfigAdminFile()
+			kubernetesConfigFlags.KubeConfig = &kubeAdminFile
+
+			config := lib.Config{
+				K8sVersion:      checkOptions.K8sVersion,
+				ForceDownload:   checkOptions.ForceDownload,
+				APIWalk:         checkOptions.APIWalk,
+				SwaggerDir:      checkOptions.SwaggerDir,
+				ShowDescription: checkOptions.ShowDescription,
+				ConfigFlags:     kubernetesConfigFlags,
+				Input:           inputFile,
+			}
+
+			klog.Infof("Starting Kubepug with configs: %+v", config)
+			kubepug := lib.NewKubepug(config)
+
+			result, err := kubepug.GetDeprecated()
+			if err != nil {
+				return err
+			}
+
+			klog.Info("Starting deprecated objects printing")
+			formatter := formatter.NewFormatter(format)
+			bytes, err := formatter.Output(*result)
+			if err != nil {
+				return err
+			}
+
+			if filename != "" {
+				err = ioutil.WriteFile(filename, bytes, 0644)
+				if err != nil {
+					return err
+				}
+			} else {
+				fmt.Printf("%s", string(bytes))
+			}
+
+			return nil
+		},
+		Args: cobra.MinimumNArgs(1),
+	}
+	cmd.PersistentFlags().BoolVar(&checkOptions.APIWalk, "api-walk", true, "Wether to walk in the whole API, checking if all objects type still exists in the current swagger.json. May be IO intensive to APIServer. Defaults to true")
+	cmd.PersistentFlags().BoolVar(&checkOptions.ShowDescription, "description", true, "Wether to show the description of the deprecated object. The description may contain the solution for the deprecation. Defaults to true")
+	cmd.PersistentFlags().StringVar(&checkOptions.K8sVersion, "k8s-version", "master", "Which kubernetes release version (https://github.com/kubernetes/kubernetes/releases) should be used to validate objects. Defaults to master")
```

Comment:
Is it able to validate the admin is going to upgraded Kubernetes version?

Like the cluster currently use version `1.17.4` and is going upgrade to `1.18.2`, then skuba should auto-fills the version to `1.18.2`, not let the admin specified which version to validated.

Reply:
Agreed

Reply:
I have put some logic for it to choose the next version to be upgraded to, but it can be specified.